core/src/main/scala/org/http4s/internal/package.scala

```diff
@@ -203,4 +207,73 @@ package object internal {
     }
     h
   }
+
+  def decode[F[_]: RaiseThrowable](charset: Charset): Pipe[F, Byte, String] = { in =>
+    val decoder = charset.nioCharset.newDecoder
+    val byteBufferSize = 16
+    val byteBuffer = ByteBuffer.allocate(byteBufferSize)
+    val charBuffer =
+      CharBuffer.allocate(math.ceil(byteBufferSize.toDouble * decoder.averageCharsPerByte).toInt)
+
+    def skipByteOrderMark(chunk: Chunk[Byte]): Chunk[Byte] =
+      if (chunk.size >= 3 && chunk.take(3) == utf8Bom)
+        chunk.drop(3)
+      else chunk
+
+    def out = {
+      val s = charBuffer.flip().toString()
+      charBuffer.clear()
+      if (s.isEmpty) Pull.done else Pull.output1(s)
+    }
+
+    Pull
+      .loop[F, String, Stream[F, Byte]] { stream =>
+        stream.pull.unconsN(byteBuffer.remaining(), allowFewer = true).flatMap {
+          case None =>
+            byteBuffer.flip()
+            val result = decoder.decode(byteBuffer, charBuffer, true)
+            byteBuffer.compact()
+            result match {
+              case _ if result.isUnderflow =>
+                def flushLoop: Pull[F, String, Unit] =
+                  decoder.flush(charBuffer) match {
+                    case result if result.isUnderflow =>
+                      out
+                    case result if result.isOverflow =>
+                      out >> flushLoop
+                  }
+                flushLoop.as(None)
+              case _ if result.isOverflow =>
+                out.as(Some(Stream.empty))
+              case _ if result.isMalformed =>
+                Pull.raiseError(new MalformedInputException(result.length()))
+              case _ if result.isUnmappable =>
+                Pull.raiseError(new UnmappableCharacterException(result.length()))
+            }
+          case Some((chunk, stream)) =>
+            val chunkWithoutBom = skipByteOrderMark(chunk)
```

Comment:
is this run on every chunk? or am I reading this wrong?

Reply:
It is, and that's a pre-existing defect, which we were reminded of on the original ticket.  I haven't had time to clean that up, but I wouldn't mind merging this while someone finds time to fix that.